src/vm/analysis/arithmetic_checker/tests.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright (C) 2013-2020 Blocstack PBC, a public benefit corporation
+// Copyright (C) 2020 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use chainstate::stacks::boot::BOOT_CODE_COSTS;
+use vm::analysis::{
+    arithmetic_checker::ArithmeticOnlyChecker, arithmetic_checker::Error,
+    arithmetic_checker::Error::*, mem_type_check, ContractAnalysis,
+};
+use vm::ast::parse;
+use vm::costs::LimitedCostTracker;
+use vm::database::MemoryBackingStore;
+use vm::functions::define::DefineFunctions;
+use vm::functions::NativeFunctions;
+use vm::types::QualifiedContractIdentifier;
+use vm::variables::NativeVariables;
+
+fn arithmetic_check(contract: &str) -> Result<(), Error> {
+    let contract_identifier = QualifiedContractIdentifier::transient();
+    let expressions = parse(&contract_identifier, contract).unwrap();
+
+    let analysis = ContractAnalysis::new(
+        contract_identifier,
+        expressions,
+        LimitedCostTracker::new_free(),
+    );
+
+    ArithmeticOnlyChecker::run(&analysis)
+}
+
+fn check_good(contract: &str) {
+    let analysis = mem_type_check(contract).unwrap().1;
+    ArithmeticOnlyChecker::run(&analysis).expect("Should pass arithmetic checks");
+}
+
+#[test]
+fn test_boot_definitions() {
+    check_good(BOOT_CODE_COSTS);
+}
+
+#[test]
+fn test_bad_defines() {
+    let tests = [
+        ("(define-public (foo) (ok 1))", DefineTypeForbidden(DefineFunctions::PublicFunction)),
+        ("(define-map foo-map ((a uint)) ((b uint))) (define-private (foo) (map-get? foo-map {a: u1}))", DefineTypeForbidden(DefineFunctions::Map)),
+        ("(define-data-var foo-var uint u1) (define-private (foo) (var-get foo-var))", DefineTypeForbidden(DefineFunctions::PersistedVariable)),
+        ("(define-fungible-token tokaroos u500)", DefineTypeForbidden(DefineFunctions::FungibleToken)),
+        ("(define-fungible-token tokaroos)", DefineTypeForbidden(DefineFunctions::FungibleToken)),
+        ("(define-non-fungible-token tokaroos uint)", DefineTypeForbidden(DefineFunctions::NonFungibleToken)),
+        ("(define-trait foo-trait ((foo (uint)) (response uint uint)))", DefineTypeForbidden(DefineFunctions::Trait)),
```

Comment:
What are our expectations for top level `let` / `begin` expressions?

Reply:
As written now, they would be allowed, but still must be "arithmetic only". I can't think of a good use case for why a contract would want to do this, but also don't think there's much of a downside for the allowing it.